## API homework

**Description**

Several tasks have been completed in the Notebook:
 - working with the list of the 250 best IMDb movies (parsing the html page);
 - writing the telegram_logger decorator, which logs the launch of the declared functions and sends messages to Telegram.

**genscan_api.py**

Python API for the service http://hollywood.mit.edu/GENSCAN.html

It is able to find and cut introns in the transmitted nucleotide sequence.

You can import the following function:

- run_genscan function — executes a request similar to filling out a form on the site. Accepts as input all parameters that can be specified on the website (except Print options). Sequence is a sequence in the form of a string, sequence_file is the path to a file with a sequence that can be downloaded and used instead of sequence. The function returns an object of the GenscanOutput type (see below).

Parameters:
- sequence (default None) - Either this parameter or "sequence_file" must be indicated, 
- sequence_file (default None) - Either this parameter or "sequence" must be indicated, 
- organization (default "Vertebrate"), 
- exon_cutoff (default 1.00), 
- sequence_name (default ""). 

GenscanOutput dataclass has the following fields:

- status — the status of the cds_list request;
- a list of predicted protein sequences,
- a list of predicted cds sequences,
- intron_list — a list of predicted introns (its sequence number, its beginning and end)
 - exon_list — everything is similar to introns, but only with exons (Suboptimal exons are not shown (optimal only!!!)).


The module may be run as script with the following parameters:
- '--sequence'(DNA sequence (upper or lower case, spaces/numbers ignored). Either this parameter or "sequence_file" must be indicated),
                        type=str)
- '--sequence_file' (Path to file containing the sequence (upper or lower case, spaces/numbers ignored). Either this parameter or "sequence" must be indicated')
- '--organism' (choices=["Vertebrate", "Arabidopsis", "Maize"], Select the organism that is the DNA sequence source. Default: Vertebrate')
- '--exon_cutoff' (choices=['0.01', '0.02', '0.05', '0.10', '0.25', '0.50', '1.00'], The probability cutoff used to determine which potential exons qualify as suboptimal exons. Default 1.00')
- '--sequence_name' (Name of the DNA sequence (optional)')
- '--output_file' (Path to output file', default="genscan_output.txt")
    args = parser.parse_args()


**Requirements**

python3>=3.6

Run:
~~~sh
pip install -r requirements.txt
~~~

**Usage**

Create and activate virtual environment:
~~~sh
python3 -m venv my_venv
source my_venv/bin/activate
~~~

Run:
~~~sh
genscan_api.py --sequence TTTCAGTGGCTGCTACTCGGCGCTTCAGTCGCGGTCGCTTCAGTCGTCAGCATGGCTCGCTCGGTGACCCTGGTCTTTCTGGTGCTTGTCTCACTGACCGGCCTGTATGCTATCCAGAGTGAGTGCCTCTTTCCCCTCTCTTGGCATTAAATTTTTAGTTCTCCTTAGTTCTCCTTCCCGACGGGATTGGCCAGGGCTGCGTCTCTCGGGAAAGGAGTGGGCGTCCCGCTGCTTACAGCTTTGAGGCTACAGGGTGGATGCGCCTGTTTGCGAGTGTGCTGACGGTCAGCGCTGAAGATGGGGAAGGCTTCTCTTTTTCTCCTCTGCTGGCGGGAGCCCCTAGACTCCCTGAGCGCAAAAGGAGGGTGGGGCTGCCTGGAAAGTGCAGGGTGGAGGGCTATGGGCACGCGGAAGCTGAGCCCCGACCACGGGAAGTCAGACCGCCGCTGCGCAGCTTCATTTGCACTCGCAGAGAGGCTCGATCCAACCGAAGTCGAGAGAAAATGGGCTTGGGTCCCAGACTCTGCGATGTTTCCAAAGCTTTCTGATCATTAACTTTGTATCTAGTAATAACCTTAAAGGTCGCCGGGCAGTGGTGGCTCACGCCTTTCATGCCAGCACTTGGGAGGCAGAGGTAGGCGGATTTCTGAATTCGAGGCCAGCCTGGTCTACAGAGTGAGTTCCAGGACAGCCAGGGTTACACAAAGAAACCCTGTCTCGAAAAAATAAATAAAGGTCTCTGGGTATGAGGCTTATTGCAATGCTGAGGACCTTGTGAGCTTCTTAAGTTTGAAAGTACGTCTACAATGTAAATGTTATGGTTCCCCGTGTATTAAGACAGGGAACTAGAGGAGGCATCGAGATTTAAAGTAACTTATTCACACGATAGACTAATGGCAGTCGGAGCTGGAATTGAGGCCGATCAATGCCCAGGCCTTTGGACTGTAACCGCCATTTGTTTCTAGAAGTGTCTAGGCCTTCCTGGGTCATGGTCTGTGAAGCAGTCTGAGATGCCTTGGACCCTTGGTACCTCACACAGCGCTTCCTTTTTGGCACACTGCCTGGTTCTTTCCGCGATAGAGCCTCTGCTTTCAGTTTTGAGACAACTAAGACTTTGCCTTAAGTCAAGGTGGTTATGAACTCTGGAGTGAGGATTATTTCAGTGTGGTGACCAAGACTCGTGAGGATAACTAGTAACTACCAAAGTCGTGGAGGTAGAAATATGGCAGAGACCAGATTTAAACCTGACCCAACACTGGACTTAATTGTGGAAGGGATGGTTCTTGTCCAGAGAATTCCAGGCATGATGAACCTGTGGCATTTAGCATGTTCTTGTTTCCAAAACCTCAAATGCAAGGAAAAATGTGGTTTATGCTAATTTAAATTTATGGGGACAAAAAGAATTCAAAGCTTTACTGAGACCTTCATTCTAGCTGGCCACAGTTTTCAGGTTTTCCTCTCCCTACTGCCTTCCATGGGTGAATAAGAGGCTGCTCTTCAGAGACCGTGCACACTGAATAGATGTGTATCTGGGAGTATATTATGGCTGGGCATAGTTGTGCACATCTATAATCTCAGTATTTAGGAGACTGAGGCAGGTGGATCTCTGAATTTGAGACCAACCTGATCTACAAAGTGCATCCAGGACAGAAAAACGCTGTCTTGAAAAACCAAAGAAAGAAAGAAAGAAAGAAAAGAAAGTGTTTTATAAGCCAAGTGTGTTGAGTCATGCTGTGACCTCAGAGCTGGGGAGGCTGAGACAGGCGTTCAAGGCCAGCCTGGAAGACAGTCAGACTGTATCTCAAATTAAATAATAAGCATATAAGTAAATGAAGGGGAAAACACAAAAAGAGCCAGTGAAATGGCTCAGCAGGTACAGGCGCTTACCACTAAGCCTGAGTGTATCTCAGATCCCACTTGGTGGACGGAGAGAACTGACTTTTAAATCCACACATGAGCTACGTTTCGTGCATGTGCACACACACACACACACACACACACACACACACACACACATATATACTCTCTCAGTAAAACAATAATTTTTTTAAGAAAGTAGATATATACATTAGTGATACCTTGGTATCAGATCTGCATATTCTACACAGTAGTACTCCTTGTCATGTTGGTTGAGAAGCAGAAACAGAGTTCAAAGATGGCAGTGGTATGGCTCAGCAGCTAAAAGTACCTGCCATGCAACCTGACCACCTGAGTTCATAGAACCCACGACAGAAGGAGGAGAAAACTGTCTCCTGAAAGCCATCCAATGTCTTCACAAGTACACCTTTGGGAACTTGACTGTGTGTATGTGTGTGTGCACATACAAATAACAATGGAAAGAGTTTACATTATTTTCTTTCCATGTCCCTATAGCGTGCAGTAGCTATTCCCAACATGTTGAGCGACTTCAGCTATCTGAAAGCCCCATATCCCCTTGCACACACTGCTTACTTCCAGCAGAGTGCATTCTCAATTGTCATGGTCCTCACATCTCCCTCTGGTCCTTTACCCATCCCAACTCAGTTGCTTAAAGCTTTCCTAGCTTCTGTGGTATTTTAAATTTTGTTTGTTTTCTAGTTCATGAAACCAGAGGTTTGGTTTCAGAATGCAAACTCTGGGCTGGAACAGAGAAACCCTGGGTTGACTCGCATAGGAACCTCATTAGGGAGGAGCCAATGCTAACACCTGCCACCTGAGGGGTAATTGCTCAGCTCTCAGCACTGGATCAGACATATGTGTTGGGAAGTCTAGGGAGGAGCAATAAGCAAAAACGAAAGGGGATGAAAATAAGAAAGAAAAGTGAGAGGGGCTGACACAGCGCACAAGCACTGCGGGCTTTGATGTGGATACTGTGAAGGGTGTGCAGAATGGGATGTGACGTTTGGAAAGTTGGTGGGATTTATTAAGGGAAGGGAGGGGCACAGTTCTTGAGAGCTTCCAATAATAAAAGCTCCTTAAAAATCCACTGACAGAAGACACTGCTAAAAGCCAGGTAGGGAAAACAGAAGGTACTCGTAGGATTTTGAGGAAAAGATACTACGTTTTCAAAATGTGGGTAGACTTTGGGGGAAGCAGATCACTTATCCAGAGTAGAAATGGAACAGGGAGAAATAGAGGAACAAATGTAAGATGGTGCACGGTGCAGACTGAGCTCTGTTTTCATCTGTCTTCCCCTGTGGCCCTCAGAAACCCCTCAAATTCAAGTATACTCACGCCACCCACCGGAGAATGGGAAGCCGAACATACTGAACTGCTACGTAACACAGTTCCACCCGCCTCACATTGAAATCCAAATGCTGAAGAACGGGAAAAAAATTCCTAAAGTAGAGATGTCAGATATGTCCTTCAGCAAGGACTGGTCTTTCTATATCCTGGCTCACACTGAATTCACCCCCACTGAGACTGATACATACGCCTGCAGAGTTAAGCATGCCAGTATGGCCGAGCCCAAGACCGTCTACTGGGGTAAGCCTCAAGTTCTTCCTTACTTTCTGGACGCTCCATCTGTGTGGACTTAAAACTGCTTTGCTATTTTAAAAACCTGCATACTGAGATTGTTAGAGAATACCCCCAAAAGACTGATCTGTGGTGCCAGCAGAGACTTACACAGCAGGCCAGGAACCAGCAGTTCCCTACCCACTGTCCCAGACTGCTTCCCTCTTTCCTGAGCTTCCTCCAGCTTTCCTGGCAGCCTGGTGCATACTAAGTGTCAAAGGGCATGCCAGGTAGGACAGGCCTTGGTCTGTAACCCCTGCCTTCAGGTACCTACAATTCTGGACTGTTCCAGGGCCTTGGCAAGAAGCTAGTAGACAGGAGTGGATCTCTGGAAAGTTCTCAACAAGTACTCAAGGCTAGCAACAGCAACCTAGTCTATCTGCCTTACTGCTTACAACCCCTGGATGTCTGTTAATATGTCGTAGTATAAAAGTAAAGTTTTGTCTCCATTTTCTTTTTTCATAGATCGAGACATGTGATCAAGCATCATGATGGTAGGTTTCTTGACTTTAAATAATGGGCCTTTGGGGGTGGGGTGGAGCTGGCGGTAGAGCACATGCCTAGCGTGCATAAGGCCTTGAGTCCTGGTCCCAGCACTGCAGGAAGAGAATTGTTTGATTCACTATCCTAGGAGCTGTTTATAGACAGCTCAAACATGATAAGCATCACTGTATGAAAGACATTAATTAGGGTGGCTTGATCAGCTACAGAAGGATCCTTTGGGGTCACATTCTTTTATCCTGTGGACTGGCAGGAAGGAGGAACGTAGCCATGTCACTGGCCCTCTAAAGGGAGAACAGCACCTTTGTGCATGCAGCAAGGCAGGTAGGCAGGTAGGCAGGTAGGCAGGCGGGTGGTCAGTTACACAGCCTGCATGGGTTTGCCGACCTTCTGTTTTAACCCTGCCAGTTCTGCTGAGAAGGCCTGGGACGATGAGCTTGCTGTCTTTGTACAGCAGTGGCCTTGCTCCTCACCCAGAGGAGCCGAGTGACAGAGCTCTGCGGGTACATCTTAGCCCTTTCAGTCTCTAGTTGGTAGGCACTAGATTTACCTTCTGGAGGCTTCCGGACACTCAGGGAAAGAAATGGACAGGGTTGTAACTTCATGTAAGGCACCGTCACTGATGTGTCAGAAGGAAGTTGAGGAGCGTGAGAGGGAACGTGGGTGTCTCTGTCAGGTGGAGTCTAGTGGTAGAAAATCCAGCTTTTCGGTGAAATCCAGGGCCCTTGAGGCCAAAAGCTCACTCAAATCTTGTATTATATTTATTTCTAGGAAAGGGGAATTGATGAAGGGGGTGGGGATGGGTGATCTGCCCAGACAAGCAGTTACCAAATGGTGGAGCCAGAACTCACCTCCTTTAAGAAGGTGATTGTCCTACAAACTCATTTGATGGTGAGGTCTGGAATGTAAATAATGTAATGTTCGGCTAACCTTCTCCTTCTCTCCCTCCCTTCCCCTTTTCTTTTCAGCTCTGAAGATTCATTTGAACCTGCTTAATTACAAATCCAGTTTCTAATATGCTATACAATTTATGCACGCAGAAAGAAATAGCAATGTACACATCACCTTCTTTATATCTTACTTTAAATATTTTATGCATGTTTTCAAAAAATTGGAAATATCCTAGATAGCTGAGCAATAAATCTTCAATAAGTATTTTGATCAGAATAATAAATATAATTTTAAGAACAATAGTTGATCATATGCCAAACCCTCTGTACTTCTCATTACTTGGATGCAGTTACTCATCTTTGGTCTATCACAACATAAGTGACATACTTTCCTTTTGGTAAAGCAAAGAGGCCTAATTGAAGTCTGTCACTGTGCCCAATGCTTAGCAATTCTCACCCCCAACCCTGTGGCTACTTCTGCTTTTGTTACTTTTACTAAAAATAAAAAAC --output_file data/genscan_output_from_string.txt
~~~

~~~sh
python3 genscan_api.py --sequence_file data/test_seq.txt --output_file data/genscan_output_from_file.txt
~~~
